pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/FunctionAssignmentTailer.java

```diff
@@ -25,95 +25,157 @@
+    private Thread tailerThread;
 
-    private final Thread tailerThread;
+    @Getter
+    private MessageId lastMessageId = null;
     
     public FunctionAssignmentTailer(
             FunctionRuntimeManager functionRuntimeManager,
             ReaderBuilder readerBuilder,
             WorkerConfig workerConfig,
-            ErrorNotifier errorNotifier) throws PulsarClientException {
+            ErrorNotifier errorNotifier) {
         this.functionRuntimeManager = functionRuntimeManager;
-        
-        this.reader = readerBuilder
-          .subscriptionRolePrefix(workerConfig.getWorkerId() + "-function-runtime-manager")
-          .readerName(workerConfig.getWorkerId() + "-function-runtime-manager")
-          .topic(workerConfig.getFunctionAssignmentTopic())
-          .readCompacted(true)
-          .startMessageId(MessageId.earliest)
-          .create();
-        
-        this.tailerThread = new Thread(() -> {
-            while(isRunning) {
-                try {
-                    Message<byte[]> msg = reader.readNext();
-                    processAssignment(msg);
-                } catch (Throwable th) {
-                    if (isRunning) {
-                        log.error("Encountered error in assignment tailer", th);
-                        // trigger fatal error
-                        isRunning = false;
-                        errorNotifier.triggerError(th);
-                    } else {
-                        if (!(th instanceof InterruptedException || th.getCause() instanceof InterruptedException)) {
-                            log.warn("Encountered error when assignment tailer is not running", th);
-                        }
-                    }
+        this.hasExited = new CompletableFuture<>();
+        this.readerBuilder = readerBuilder;
+        this.workerConfig = workerConfig;
+        this.errorNotifier = errorNotifier;
+    }
 
-                }
+    public synchronized CompletableFuture<Void> triggerReadToTheEndAndExit() {
+        exitOnEndOfTopic = true;
+        return this.hasExited;
+    }
+
+    public void startFromMessage(MessageId startMessageId) throws PulsarClientException {
+        log.info("Function assignment tailer start reading from topic {} at {}",
+                workerConfig.getFunctionAssignmentTopic(), startMessageId);
+        if (!isRunning) {
+            isRunning = true;
+            if (reader == null) {
+                reader = createReader(startMessageId);
             }
-        });
-        this.tailerThread.setName("assignment-tailer-thread");
+            if (tailerThread == null || !tailerThread.isAlive()) {
+                tailerThread = getTailerThread();
+            }
+            hasExited = new CompletableFuture<>();
+            tailerThread.start();
+        }
     }
 
-    public void start() {
-        isRunning = true;
-        tailerThread.start();
+    public synchronized void start() throws PulsarClientException {
```

Comment:
I think its cleaner to consolidate this and above method to start(MessageId) { ... }

Reply:
I also think that some logic will be simpler if we create Tailer object every time we go thru leadership transistion

Reply:
> I also think that some logic will be simpler if we create Tailer object every time we go thru leadership transition

That is not correct.  The functionAssignmentTailer is also responsible for keeping track of a message id.  If a worker becomes a leader and then loses leadership prior to creating any assignments, we shouldn't just start reading the assignment topic from the beginning.  We should resume from the message id stored in the functionAssignmentTailer